src/ProxyClient/Http/HttpAdapter.php

```diff
@@ -0,0 +1,293 @@
+use FOS\HttpCache\Exception\InvalidArgumentException;
+use FOS\HttpCache\Exception\InvalidUrlException;
+use FOS\HttpCache\Exception\ProxyResponseException;
+use FOS\HttpCache\Exception\ProxyUnreachableException;
+use Http\Client\Exception;
+use Http\Client\Exception\HttpException;
+use Http\Client\Exception\RequestException;
+use Http\Client\HttpAsyncClient;
+use Http\Client\Promise;
+use Http\Discovery\UriFactoryDiscovery;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\UriInterface;
+
+/**
+ * An adapter to work with the Httplug asynchronous client.
+ *
+ * @author David Buchmann <mail@davidbu.ch>
+ */
+class HttpAdapter
+{
+    /**
+     * @var HttpAsyncClient
+     */
+    private $httpClient;
+
+    /**
+     * Asynchronously sent requests
+     *
+     * @var Promise[]
+     */
+    private $queue = [];
+
+    /**
+     * Caching proxy server host names or IP addresses.
+     *
+     * @var UriInterface[]
+     */
+    private $servers;
+
+    /**
+     * Application host name and optional base URL.
+     *
+     * @var UriInterface
+     */
+    private $baseUri;
+
+    /**
+     * @param string[] $servers Caching proxy server hostnames or IP
+     *                          addresses, including port if not port 80.
+     *                          E.g. ['127.0.0.1:6081']
+     * @param string   $baseUri Default application hostname, optionally
+     *                          including base URL, for purge and refresh
+     *                          requests (optional). This is required if
+     *                          you purge and refresh paths instead of
+     *                          absolute URLs.
+     * @param HttpAsyncClient $httpClient
+     */
+    public function __construct(array $servers, $baseUri, HttpAsyncClient $httpClient)
+    {
+        $this->setServers($servers);
+        $this->setBaseUri($baseUri);
+        $this->httpClient = $httpClient;
+    }
+
+    /**
+     * Queue invalidation request.
+     *
+     * @param RequestInterface $invalidationRequest
+     */
+    public function invalidate(RequestInterface $invalidationRequest)
+    {
+        $signature = $this->getRequestSignature($invalidationRequest);
+
+        if (isset($this->queue[$signature])) {
+            return;
+        }
+
+        $this->queue[$signature] = $invalidationRequest;
+    }
+
+    /**
+     * Send all pending invalidation requests and make sure the requests have terminated and gather exceptions.
+     *
+     * @return int The number of cache invalidations performed per caching server.
+     *
+     * @throws ExceptionCollection If any errors occurred during flush.
+     */
+    public function flush()
+    {
+        $queue = $this->queue;
+        $this->queue = [];
+        /** @var Promise[] $promises */
+        $promises = [];
+
+        $exceptions = new ExceptionCollection();
+
+        foreach ($queue as $request) {
+            foreach ($this->multiplex($request) as $proxyRequest) {
```

Comment:
I’m not sure ‘multiplex’ is the right terminology. Perhaps what you mean is [inverse multiplexing](https://en.wikipedia.org/wiki/Inverse_multiplexing)? Unfortunately, that sounds a little convoluted, so perhaps just `fanOut()` or something would be better?